CFD-320 : Homepage : Activity stream should contain only trimmed text

nader77 opened the pull request:
#526

nader77 commented:
Trimmed activity stream on the homepage:
![selection_073](https://cloud.githubusercontent.com/assets/7369740/6430066/52271f14-c002-11e4-98e0-862e85f9aaa8.png)

nader77 commented:
@zero2one PR is ready.

nader77 commented:
Update: This PR does the following
* Trim body content to 150 chars, strip tags and add '...' at the end in the activity stream of the homepage.
* Fix the parameters that determine  the content of the activity stream in each group.
* Hide "Show more" button fix.

amitaibu commented on capacity4more/modules/c4m/message/c4m_message/c4m_message.module:
```diff
@@ -187,7 +187,7 @@ function c4m_messages_group_info(Message $message) {
   // When the request is from RESTful, "drupal_is_front_page" won't work.
   $handler = og_purl_og_context_handler();
-  if ((drupal_is_front_page() && empty($handler) || (isset($_REQUEST['group']) && is_array($_REQUEST['group'])))) {
+  if (((drupal_is_front_page() || !empty($_REQUEST['homepage'])) && empty($handler) || (isset($_REQUEST['group']) && is_array($_REQUEST['group'])))) {
```
``$_REQUEST`` should be ``$_GET``, no?

nader77 commented:
@amitaibu Comments fixed.

zero2one commented:
@amitaibu @nader77 

I'm sorry but I'm not gonna approve this solution.
We need to avoid hard-coded display of the entities in the activity stream.

* Drupal has display modes for entities, we should use them. 
* The settings can be featured in the content type features.
* We use Display suite to define the display of the display modes.
* We should use the same approach everywhere. This makes it easier to maintain.

### Message entity display modes
2 different display modes:
* global activity
* group activity

Where 
* "global activity" uses the "global activity" display mode of the node entity to render the node teaser. 
* "group activity" uses the "group activity" display mode of the node entity to render the node teaser.

### Node entity display modes
2 different display modes:
* global activity : used in the activity streams outside groups (homepage).
* group activity : used in the activity streams inside groups (group dashboard).

zero2one commented:
@amitaibu @nader77 

Parked for next sprint.

zero2one commented:
See CFD-352

nader77 commented:
Update:
* Created 2 display modes ``activity_group`` && ``activity_global``
* Updated all the content features with the new display modes.
* Added the module [Smart trim](https://www.drupal.org/project/smart_trim) which gives us dynamic trimming, suffex for the trimming and a strip tags option:
![selection_062](https://cloud.githubusercontent.com/assets/7369740/6769781/1f3842d0-d0b1-11e4-8ec9-cab6cc2bb270.png)

nader77 commented:
Homepage:
![selection_064](https://cloud.githubusercontent.com/assets/7369740/6769795/7d409cb0-d0b1-11e4-9072-0a387c152eac.png)

Group dashboard: 
![selection_065](https://cloud.githubusercontent.com/assets/7369740/6769796/8318335a-d0b1-11e4-8355-fa1ec6ca2973.png)

nader77 commented:
@HelenaEksler PR is ready for code review.

HelenaEksler commented:
@nader77 Looks good for me.
@zero2one PR is ready for your review.

zero2one commented:
Can you do following changes:

## Change code
* Remove outputting the node title directly to the activity stream in code.
* Remove outputting the number of comments & views from the code.

## Add the node title activity display modes
* Add the node title to the activity display modes (Activity Group & Activity Global).
* Add the "Footer stats" (number of comments & views) to the activity display modes.
* Update css if necessary.

## How to add the title field & footer stats to the node display:

Open the display mode setting for the node you want to update.
Select a layout
![c4m-display-01](https://cloud.githubusercontent.com/assets/133124/6824722/00f07360-d2f5-11e4-9957-18e2e3a1f047.png)

Choose for "One column".
![c4m-display-02](https://cloud.githubusercontent.com/assets/133124/6824799/928bc004-d2f5-11e4-87b6-60001a2503ba.png)

Save the settings
![c4m-display-03](https://cloud.githubusercontent.com/assets/133124/6824802/9ac32f1e-d2f5-11e4-9636-e04798e12763.png)

Once the settings are saved, a lot more fields will be available to add to the display.
![c4m-display-04](https://cloud.githubusercontent.com/assets/133124/6824807/a10b81c8-d2f5-11e4-9321-962839a1b0d6.png)

Drag and drop the title field above the body.
Drag and drop the footer stats underneath the body.
![c4m-display-05](https://cloud.githubusercontent.com/assets/133124/6824903/8cee8bd0-d2f6-11e4-82fd-f41b18784d4b.png)

Save the settings.

nader77 commented:
@zero2one Before i do the changes, Just wanna mention that the title is created in the message not in the code itself:
![edit_c4m_insert_comment___capacity4more](https://cloud.githubusercontent.com/assets/7369740/6841934/4188f08c-d39a-11e4-97d4-981a77c41f53.png)

Does this change anything or i should still delete it from here and create it in the display mode ?

nader77 commented:
Footer is created in the code.

zero2one commented:
@nader77 

I know but we realy should use the rendered node.

It makes it clearer for everybody that the node is rendered with a certain display mode + the display of the node part is managed in the node iteself.

nader77 commented:
@zero2one No problem, Just making sure :)

nader77 commented:
Update:
Added title and footer stats to all content types, Removed from code:
![article___capacity4more](https://cloud.githubusercontent.com/assets/7369740/6844276/de53fc64-d3b2-11e4-9c98-59b546b43d8f.png)

In comments, I had to keep the title from the message because when i tried to create view modes and add the node's title, I found out that i have only the comment title, Couldn't add the node's title:
![article___capacity4more](https://cloud.githubusercontent.com/assets/7369740/6844285/f2bbf3dc-d3b2-11e4-9d8b-f7020c140d9a.png)

nader77 commented:
@zero2one @HelenaEksler 
Activity stream after all the changes, Please point out if there's some unwanted changes or errors, Thanks.

![activity_stream](https://cloud.githubusercontent.com/assets/7369740/6846839/2957152c-d3ca-11e4-9985-8a8df06515db.gif)
